### PR TITLE
Build System pt.4 - Persistence, Surface alignment, gui improvements

### PR DESCRIPTION
**Pull request by veteran29**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| Needs wipe? | no |
| Fixed issues | #507, #505  |

### Description:
Further development of build system

### Content:
- [x] Align objects to surface
- [x] Show objects target position ~~bounding box~~, changed to show icon at target area
- [x] Allow to place objects on surfaces like roofs
- [x] Add icons for dragging
- [x] Allow to build only in FOB area 
- [x] Fix boats spawning at bottom of water containers
- [x] Object persistence

### Tested on:
- [x] Local MP Vanilla
- [x] ~~Local MP ACE~~
- [x] Dedicated MP Vanilla
- [x] ~~Dedicated MP ACE~~

### CODE REVIEW

File: Missionframework/modules/04_build/fnc/fn_build_handleKeys.sqf [hunk 1]
```diff
@@ -26,6 +26,10 @@ params [
 switch toLower _mode do {
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_handleKeys.sqf]: Header -> Parameters

File: Missionframework/modules/04_build/fnc/fn_build_handleMouse.sqf [hunk 2]
```diff
@@ -6,18 +6,18 @@
...
 
     Description:
-    Handle build display mouse movement events
+        Handle build display mouse movement events
 
     Parameter(s):
-        0: STRING - Mouse move event
-        1: ARRAY - Additional event arguments
+        _mode   - Mouse event name              [STRING]
+        _args   - Additonal event parameters    [ANY]
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_handleMouse.sqf]: ARRAY, defaults to nil

_args just accepts an array, so `["myMode", "myArgs"] call KPLIB_fnc_build_handleMouse;` would (should) create an error.

File: Missionframework/modules/04_build/fnc/fn_build_handleRotation.sqf [hunk 3]
```diff
@@ -5,17 +5,18 @@
...
+        Handles object rotation dragging
 
     Parameter(s):
-    NONE
+        _anchorObject   - Rotation dragging anchor object                       [OBJECT, defaults to objNull]
+        _updatePos      - Rotation should be finished and positions updated     [BOOL, defaults to false]
 
     Returns:
-    NOTHING
+        NOTHING
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_handleRotation.sqf]: Despite it has 2 exitWith, we should try to return BOOL. Or is there anything which I'm missing in this particular function?

File: Missionframework/modules/04_build/fnc/fn_build_loadData.sqf [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+if (KPLIB_param_debugSave > 0) then {diag_log "[KP LIBERATION] [SAVE] Build module loading...";};
+
+private _moduleData = ["build"] call KPLIB_fnc_init_getSaveData;
+
+// Check if there is a new campaign
+if (_moduleData isEqualTo []) then {
+    if (KPLIB_param_debugSave > 0) then {diag_log "[KP LIBERATION] [SAVE] Build module data empty, creating new data...";};
+
+    // Initialize every sector and add the data to the garrison array
+    KPLIB_build_save_data = [] call CBA_fnc_createNamespace;
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_loadData.sqf]:
```suggestion
    KPLIB_build_saveData = [] call CBA_fnc_createNamespace;
```

or

```suggestion
    KPLIB_build_saveNamespace = [] call CBA_fnc_createNamespace;
```

File: Missionframework/modules/04_build/fnc/fn_build_markArea.sqf [hunk 5]
```diff
@@ -4,18 +4,18 @@
...
 
     Description:
-    Creates sphere indicators around given area (LOCALLY)
+        Creates sphere indicators around given area (LOCALLY)
 
     Parameter(s):
-        0: ARRAY - Center position of area to mark
-        1: STRING - Radius to mark around the center position
+        _position   - Center position of area to mark           [ARRAY]
+        _radius     - Radius to mark around the center position [NUMBER]
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_markArea.sqf]: defaults to nil

File: Missionframework/modules/04_build/fnc/fn_build_markArea.sqf [hunk 6]
```diff
@@ -4,18 +4,18 @@
...
 
     Returns:
-    ARRAY - Array with created spheres
+        Array with created spheres [ARRAY]
 */
 params [
     ["_position", nil, [[]]],
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_markArea.sqf]:
```suggestion
    ["_position", nil, [[]], 3],
```

File: Missionframework/modules/04_build/fnc/fn_build_preInit.sqf [hunk 7]
```diff
@@ -31,6 +31,30 @@ KPLIB_buildLogic = locationNull;
 // Build camera PFH ticker id
 KPLIB_build_ticker = -1;
 
+// Save data
+KPLIB_build_save_data = locationNull;
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_preInit.sqf]:
```suggestion
KPLIB_build_saveNamespace = locationNull;
```

File: Missionframework/modules/04_build/fnc/fn_build_saveData.sqf [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    // Foreach fob item, save className, pos and dirAndUp
+    {
+        if (alive _x) then {
+            (_fobData select 1) pushBack [typeOf _x, getPosATL _x, [vectorDir _x, vectorUp _x]];
+        };
+    } forEach _fobItems;
+
+    _serializedData pushBack _fobData;
+
+} forEach (allVariables KPLIB_build_save_data);
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_saveData.sqf]:
```suggestion
} forEach (allVariables KPLIB_build_saveNamespace);
```

File: Missionframework/modules/04_build/fnc/fn_build_start.sqf [hunk 9]
```diff
@@ -6,18 +6,18 @@
...
     License: GNU General Public License v3.0 - https://www.gnu.org/licenses/gpl-3.0.html
 
     Description:
     Starts KP Liberation building mode
 
     Parameter(s):
-        0: ARRAY - Center position of build area
-        1: NUMBER - Allowed build radius
+        _center - Center of building area [POSITION]
+        _radius - Allowed building radius [NUMBER]
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_start.sqf]: defaults to nil

File: Missionframework/modules/04_build/fnc/fn_build_validatePosition.sqf [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    Author: KP Liberation Dev Team - https://github.com/KillahPotatoes
+    Date: 2018-11-04
+    Last Update: 2018-11-04
+    License: GNU General Public License v3.0 - https://www.gnu.org/licenses/gpl-3.0.html
+
+    Description:
+        Validates object placement and sets variable on it accordingly
+
+    Parameter(s):
+        _object - Object to validate [OBJECT]
```
Comment (Wyqer) [inline Missionframework/modules/04_build/fnc/fn_build_validatePosition.sqf]:
```suggestion
        _object - Object to validate [OBJECT, defaults to nil]
```

File: Missionframework/modules/04_build/fnc/fn_build_addToSelection.sqf [hunk 11]
```diff
@@ -5,17 +5,17 @@
...
+    Last Update: 2018-11-04
     License: GNU General Public License v3.0 - https://www.gnu.org/licenses/gpl-3.0.html
 
     Description:
-    Adds objects to selection
+        Manges selection
 
     Parameter(s):
-        0: OBJECT/ARRAY - Object or array of objects to be added
+        _selection - Items to add to selection [Object or Array, defaults to objNull]
```
[the remark this answers is not in the document]
Comment (veteran29) [reply Missionframework/modules/04_build/fnc/fn_build_addToSelection.sqf]: Yeah, it just copies everything into the tooltip:
image

File: Missionframework/modules/01_common/fnc/fn_common_preInit.sqf [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    Author: KP Liberation Dev Team - https://github.com/KillahPotatoes
+    Date: 2018-11-04
+    Last Update: 2018-11-04
+    License: GNU General Public License v3.0 - https://www.gnu.org/licenses/gpl-3.0.html
+
+    Description:
+        This preInit function defines global variables which are added due to this module.
+
+    Dependencies:
+        * 02_core
```
[the remark this answers is not in the document]
Comment (veteran29) [reply Missionframework/modules/01_common/fnc/fn_common_preInit.sqf]: Added `01_common` as it uses getIcon function

[the remark this answers is not in the document]
Comment (Wyqer) [reply Missionframework/modules/01_common/fnc/fn_common_preInit.sqf]: it's the common preInit file ^^

[the remark this answers is not in the document]
Comment (veteran29) [reply Missionframework/modules/01_common/fnc/fn_common_preInit.sqf]: Oh I did not notice that. :|

**Pull request merged**